src/Features/Core/Portable/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -0,0 +1,243 @@
+
+            // Need to do this with index so we get the line endings correct.
+            for (var i = 0; i < trivia.Count; i++)
+            {
+                var triviaLine = trivia[i];
+                if (triviaLine.RawKind == SyntaxKinds.SingleLineCommentTrivia)
+                {
+                    if (possibleLeadingSpaces != string.Empty)
+                    {
+                        leadingSpaces = possibleLeadingSpaces;
+                    }
+
+                    removalList.Add(i);
+                    onBlankLine = false;
+                }
+                else if (triviaLine.RawKind == SyntaxKinds.WhitespaceTrivia)
+                {
+                    if (leadingSpaces == string.Empty)
+                    {
+                        possibleLeadingSpaces = triviaLine.ToFullString();
+                    }
+
+                    if (inCopyright)
+                    {
+                        removalList.Add(i);
+                    }
+                }
+                else if (triviaLine.RawKind == SyntaxKinds.EndOfLineTrivia)
+                {
+                    possibleLeadingSpaces = string.Empty;
+
+                    if (inCopyright)
+                    {
+                        removalList.Add(i);
+                    }
+
+                    if (onBlankLine)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        onBlankLine = true;
+                    }
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            // Remove copyright lines in reverse order.
+            for (var i = removalList.Count - 1; i >= 0; i--)
+            {
+                trivia = trivia.RemoveAt(removalList[i]);
+            }
+
+            var newLineText = document.Project.Solution.Options.GetOption(FormattingOptions.NewLine, root.Language)!;
+            var newLineTrivia = SyntaxGenerator.GetGenerator(document).EndOfLine(newLineText);
+
+            var newHeaderTrivia = CreateNewHeader(leadingSpaces + FileHeaderHelper.CommentPrefix, expectedFileHeader, newLineText);
+
+            // Add a blank line after the header.
+            newHeaderTrivia = newHeaderTrivia.Add(newLineTrivia);
+
+            // Insert header at top of the file.
+            return root.WithLeadingTrivia(newHeaderTrivia.Add(newLineTrivia).AddRange(trivia));
+        }
+
+        private SyntaxNode AddHeader(Document document, SyntaxNode root, string expectedFileHeader)
+        {
+            var newLineText = document.Project.Solution.Options.GetOption(FormattingOptions.NewLine, root.Language)!;
+            var newLineTrivia = SyntaxGenerator.GetGenerator(document).EndOfLine(newLineText);
+            var newTrivia = CreateNewHeader(FileHeaderHelper.CommentPrefix, expectedFileHeader, newLineText).Add(newLineTrivia).Add(newLineTrivia);
```

Comment:
so far, in both cases i've seen you call CreateNewHeader you add two newlines afterwards.  can this just move into CreateNewHeader if that always needs to happen?

Reply:
Eventually we won't want this behavior (e.g. the code fix for `TestInvalidFileHeaderWithWrongTextInRegionAsync`), so I'll leave it in the current form for now.